extensions/positron-r/src/lsp.ts

```diff
@@ -40,7 +40,7 @@ export enum LspState {
 export class ArkLsp implements vscode.Disposable {
 
 	/** The languge client instance, if it has been created */
-	private _client?: LanguageClient;
+	client?: LanguageClient;
```

Comment:
I quite strongly feel that we should not make internal details public just for tests.

I think you can do two things in your tests to get access to what you need here:

- Use the already public `onDidChangeState()` event on `ArkLsp`, you don't need the one from the `client`
- Next to the `private setState(state: LspState)` method in `ArkLsp`, add a public `getState(): LspState` method that just returns the current state, and use that in your tests.

It looks like that'll easily give you everything you need, and you can keep `private _client`, which seems like a big win to me.

Reply:
My thought was that exposing the client makes sense in general. It's only within the extension, i.e. `pub(crate)`, and there's a lot of useful stuff you can do with the client such as sending a custom notification. If we try to hide it away we could end up recreating the `LanguageClient` API on our `Lsp` type. The benefits of preventing access to the subset of methods that are really private to `Lsp` (e.g. state changes) are limited since the scope of the extension is not that large.

Reply:
I'm now using `ArkLspState`. The event now matches `LanguageClient` with `oldState` and `newState` fields. The client is now private again, I agree I went for the wrong level here.